noodles-util/src/alignment/async/io/reader/builder.rs

```diff
@@ -0,0 +1,174 @@
+    ) -> io::Result<Reader<Box<dyn AsyncBufRead + Unpin>>>
+    where
+        P: AsRef<std::path::Path>,
+    {
+        let file = File::open(src).await?;
+        self.build_from_reader(file).await
+    }
+
+    /// Builds an async alignment reader from a reader.
+    ///
+    /// By default, the format and compression method will be autodetected. This can be overridden
+    /// by using [`Self::set_format`] and [`Self::set_compression_method`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[tokio::main]
+    /// # async fn main() -> tokio::io::Result<()> {
+    /// use noodles_util::alignment::async::io::reader::Builder;
+    /// use tokio::io;
+    /// let reader = Builder::default().build_from_reader(io::empty()).await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn build_from_reader<R>(
+        self,
+        reader: R,
+    ) -> io::Result<Reader<Box<dyn AsyncBufRead + Unpin>>>
+    where
+        R: AsyncRead + Unpin + 'static,
+    {
+        use crate::alignment::io::reader::builder::{detect_compression_method, detect_format};
+        let mut reader = BufReader::new(reader);
+
+        let compression_method = match self.compression_method {
+            Some(compression_method) => compression_method,
+            None => {
+                let mut src = reader.fill_buf().await?;
+                detect_compression_method(&mut src)?
+            },
+        };
+
+        let format = match self.format {
+            Some(format) => format,
+            None => {
+                let mut src = reader.fill_buf().await?;
+                detect_format(&mut src, compression_method)?
+            }
+        };
+
+        let reader: Box<dyn AsyncBufRead + Unpin> = match (format, compression_method) {
+            (Format::Sam, None) => Box::new(reader),
+            (Format::Sam, Some(CompressionMethod::Bgzf)) => Box::new(bgzf::AsyncReader::new(reader)),
+            (Format::Bam, None) => Box::new(reader),
+            (Format::Bam, Some(CompressionMethod::Bgzf)) => Box::new(bgzf::AsyncReader::new(reader)),
+            (Format::Cram, None) => {
+                let inner: Box<dyn AsyncBufRead + Unpin> = Box::new(reader);
+                let inner = cram::r#async::io::reader::Builder::default()
+                    .set_reference_sequence_repository(self.reference_sequence_repository)
+                    .build_from_reader(inner);
+                return Ok(Reader::Cram(inner));
+            }
+            (Format::Cram, Some(CompressionMethod::Bgzf)) => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "CRAM cannot be compressed with BGZF",
+                ));
+            }
+        };
+
+        let reader: Reader<Box<dyn AsyncBufRead + Unpin>> = match format {
+            Format::Sam => Reader::Sam(sam::r#async::io::Reader::new(reader)),
+            Format::Bam => Reader::Bam(bam::r#async::io::Reader::new(reader)),
```

Comment:
```suggestion
            Format::Bam => Reader::Bam(bam::r#async::io::Reader::from(reader)),
```

[`Reader::new`](https://docs.rs/noodles/0.79.0/noodles/bam/async/io/struct.Reader.html#method.new) wraps a raw stream with a BGZF decoder. Since you already have a decoder, you want to give that to the BAM reader directly using [`Reader::from`](https://docs.rs/noodles/0.79.0/noodles/bam/async/io/struct.Reader.html#impl-From%3CR%3E-for-AsyncReader%3CR%3E).

Reply:
Ahhhhh silly me. Thank you!